modules/token/keeper/grpc_query.go

```diff
@@ -145,24 +168,83 @@ func (k Keeper) Fees(c context.Context, req *v1.QueryFeesRequest) (*v1.QueryFees
 	}, nil
 }
 
-// Params return the all the parameter in tonken module
-func (k Keeper) Params(
-	c context.Context,
-	req *v1.QueryParamsRequest,
-) (*v1.QueryParamsResponse, error) {
+// Params returns all the parameters in the token module.
+//
+// Parameters:
+// - c: Context object
+// - req: QueryParamsRequest object
+//
+// Returns:
+// - QueryParamsResponse object containing token params
+// - Error if any
+func (k Keeper) Params(c context.Context, req *v1.QueryParamsRequest) (*v1.QueryParamsResponse, error) {
 	ctx := sdk.UnwrapSDKContext(c)
 	params := k.GetParams(ctx)
 
 	return &v1.QueryParamsResponse{Params: params}, nil
 }
 
 // TotalBurn return the all burn coin
-func (k Keeper) TotalBurn(
-	c context.Context,
-	req *v1.QueryTotalBurnRequest,
-) (*v1.QueryTotalBurnResponse, error) {
+//
+// Parameters:
+// - c: Context object
+// - req: QueryFeesRequest object
+//
+// Returns:
+// - QueryTotalBurnResponse object containing token params
+// - Error if any
+func (k Keeper) TotalBurn(c context.Context, req *v1.QueryTotalBurnRequest) (*v1.QueryTotalBurnResponse, error) {
 	ctx := sdk.UnwrapSDKContext(c)
+
 	return &v1.QueryTotalBurnResponse{
 		BurnedCoins: k.GetAllBurnCoin(ctx),
 	}, nil
 }
+
+// Balances retrieves the balances of a given address for a specific token.
+//
+// Parameters:
+// - c: the context.Context object.
+// - req: the v1.QueryBalancesRequest object containing the address and token denomination.
+//
+// Returns:
+// - *v1.QueryBalancesResponse: the response containing the balances of the address for the specified token.
+// - error: an error if the request is empty, the address is invalid, or the token is not found.
+func (k Keeper) Balances(c context.Context, req *v1.QueryBalancesRequest) (*v1.QueryBalancesResponse, error) {
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	addr, err := sdk.AccAddressFromBech32(req.Address)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid address (%s)", err)
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+
+	if !k.HasToken(ctx, req.Denom) {
+		balance := k.bankKeeper.GetBalance(ctx, addr, req.Denom)
+		balances := sdk.NewCoins(balance)
+		return &v1.QueryBalancesResponse{Balances: balances}, nil
+	}
+
+	token, err := k.GetToken(ctx, req.Denom)
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "token %s not found", req.Denom)
+	}
+
+	balance := k.bankKeeper.GetBalance(ctx, addr, token.GetMinUnit())
+	balances := sdk.NewCoins(balance)
+
+	if len(token.GetContract()) > 0 {
+		contract := common.HexToAddress(token.GetContract())
+		account := common.BytesToAddress(addr.Bytes())
+
+		erc20Balance, err := k.BalanceOf(ctx, contract, account)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		balances = balances.Add(sdk.NewCoin(token.GetContract(), sdkmath.NewIntFromBigInt(erc20Balance)))
+	}
```

Comment:
```suggestion
        balance := k.bankKeeper.GetBalance(ctx, addr, req.Denom)
        balances := sdk.NewCoins(balance)

	token, err := k.GetToken(ctx, req.Denom)
	if err != nil {
		return &v1.QueryBalancesResponse{Balances: balances}, nil
	}

	if len(token.GetContract()) > 0 {
		contract := common.HexToAddress(token.GetContract())
		account := common.BytesToAddress(addr.Bytes())

		erc20Balance, err := k.BalanceOf(ctx, contract, account)
		if err != nil {
			return nil, status.Error(codes.Internal, err.Error())
		}
		balances = balances.Add(sdk.NewCoin(token.GetContract(), sdkmath.NewIntFromBigInt(erc20Balance)))
	}
```

Reply:
There is a problem with writing this way, it will miss denom, which is the balance of the symbol.